File: facebook/OC/LC/LC/LC215.m
//
//  LC215.m
//  LC
//
//  Created by ULS on 2/18/18.
//  Copyright © 2018 ULS. All rights reserved.
//

#import "LC215.h"

@interface LC215()

- (int)findKthLargest:(int *)nums size:(int)size k:(int)k;

@end

@implementation LC215

- (int)findKthLargest:(int *)nums size:(int)size k:(int)k {
    if(size == 0) {
        return -1;
    }
    [self shuffle:nums size:size];
    k = size - k;
    int i = 0;
    int j = size - 1;
    while(i <= j) {
        int index = [self partition:nums start:i end:j];
//        DLog(@"index:%d => nums[%d]", index, nums[index]);
        if(index == k) {
            return nums[index];
        } else if (index < k) {
            i = index + 1;
        } else {
            j = index - 1;
        }
//        DLog(@"i:%d /// j:%d", i, j);
    }
    return -1;
}

- (int)partition:(int *)nums start:(int)start end:(int)end {
    int i = start;
    int j = end;
    int pivot = nums[start];
    /// DLog(@"pivot:%d", pivot);
    while(i < j) {
        while(i < end && nums[i] <= pivot) {
            i++;
        }
        while(start < j && nums[j] >= pivot) {
            j--;
        }
//        DLog(@"i:%d j:%d", i, j);
        if(i >= j) {
            break;
        }
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    int temp = nums[start];
    nums[start] = nums[j];
    nums[j] = temp;
//    for(int i = start; i <= end; i++) {
//        DLog(@"%d", nums[i]);
//    }
    return j;
}
- (void)shuffle:(int *)nums size:(int)size {
    for(int i = 1; i < size; i++) {
        int r = arc4random_uniform(i + 1);
        int temp = nums[i];
        nums[i] = nums[r];
        nums[r] = temp;
    }
}

- (void)test {
//    int nums[2] = {2, 1};
//    DLog(@"res:%d", [self findKthLargest:nums size:2 k:2]);
//    int nums[1] = {1};
//    DLog(@"res:%d", [self findKthLargest:nums size:1 k:1]);
    // int nums[6] = {6,5,3,2,1,4};
    int nums[6] = {4, 5, 3, 2, 1, 6};
    DLog(@"res:%d", [self findKthLargest:nums size:6 k:6]);
    DLog(@"res:%d", [self findKthLargest:nums size:6 k:5]);
    DLog(@"res:%d", [self findKthLargest:nums size:6 k:4]);
    DLog(@"res:%d", [self findKthLargest:nums size:6 k:3]);
    DLog(@"res:%d", [self findKthLargest:nums size:6 k:2]);
    DLog(@"res:%d", [self findKthLargest:nums size:6 k:1]);
    DLog(@"res:%d", [self findKthLargest:nums size:6 k:0]);
}

@end

File: LCIntelliJ/src/LC215.java
import java.util.Arrays;
import java.util.Random;

public class LC215 {

    public int findKthLargest_1(int[] nums, int k) {
        if(nums.length == 0) {
            return 0;
        }
        shuffle_1(nums);
        k = nums.length - k;
        int i = 0;
        int j = nums.length - 1;
        int index = -1;
        while (i <= j) {
            index = partition_1(nums, i, j);
            if(index == k) {
                break;
            } else if (index < k) {
                i = index + 1;
            } else {
                j = index - 1;
            }
        }

        if(i > j) {
            return nums[j];
        }
        return nums[index];
    }

    public int partition_1(int[] nums, int start, int end) {
        int i = start;
        int j = end;
        int pivot = nums[start];
        while (i < j) {
            while(i < end && nums[i] <= pivot) {
                i++;
            }
            while(start < j && nums[j] >= pivot) {
                j--;
            }

            if(i < j) {
                swap_1(nums, i, j);
                i++;
                j--;
            }
        }
        swap_1(nums, start, j);
        return j;
    }

    private void swap_1(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    public void shuffle_1(int[] a) {
        final Random random = new Random();
        for(int ind = 1; ind < a.length; ind++) {
            final int r = random.nextInt(ind + 1);
            swap_1(a, ind, r);
        }
    }

    public int findKthLargest(int[] nums, int k) {
        k = nums.length - k;
        shuffle(nums);
        int i = 0;
        int j = nums.length - 1;
        int index = -1;
        while(i <= j) {
            index = partition(nums, i, j);
            if(index == k) {
                break;
            } else if (index > k) {
                j = index - 1;
            } else {
                i = index + 1;
            }
        }
        if(i > j) {
            return nums[j];
        }
        return nums[index];
    }

    private int partition(int[] nums, int low, int high) {
        int i = low;
        int j = high;
        int pivot = nums[low];
        while(i < j) {
            while(i < high && nums[i] <= pivot) {
                i++;
            }
            while(j > low && nums[j] >= pivot) {
                j--;
            }
            if(i < j) {
                swamp(nums, i, j);
                i++;
                j--;
            }
        }

        swamp(nums, low, j);
        return j;
    }

    private void shuffle(int[] nums) {
        final Random random = new Random();
        for(int i = 1; i < nums.length; i++) {
            final int r = random.nextInt(i + 1);
            swamp(nums, i, r);
        }
    }

    private void swamp(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    public static void main(String[] args) {
        int[] nums = new int[]{3, 2, 3, 1, 2, 4, 5, 5, 6};
        int[] copyNums = Arrays.copyOf(nums, nums.length);
        LC215 obj = new LC215();
        int res = obj.findKthLargest_1(nums, 4);
        Arrays.sort(copyNums);
        PrintUtils.printArray(copyNums);
        PrintUtils.printString("res:" + res + " == " + copyNums[copyNums.length - 4]);

//        int[] nums = new int[]{5, 2, 4, 1, 3, 6, 0};
//        LC215 obj = new LC215();
//        int res = obj.findKthLargest_1(nums, 4);
//        PrintUtils.printString("res:" + res);
    }

}
